### ISSUE TITLE
rancher2_cluster_sync failed with [file-deployer] is still running

### ISSUE THREAD

**Issue by koboltmarky** [rancher/terraform-provider-rancher2#701]
Hi,

we have the problem that the rancher2_cluster_sync resource failed with an error.

Versions:

```
terraform 1.0.0
rancher2 = {
      source  = "rancher/rancher2"
      version = "1.15.1"
    }
    rke = {
      source  = "rancher/rke"
      version = "1.2.1"
  }
```

Resource:
```
resource "rancher2_cluster_sync" "wait-for-cluster-is-ready" {
  depends_on = [
    null_resource.init_cluster,
  ]
  cluster_id      = rancher2_cluster.rancher-cluster.id
  state_confirm   = 2
  wait_alerting   = false
  wait_catalogs   = false
  wait_monitoring = false
}

```

```
Error: Cluster ID c-5srfq: Container [file-deployer] is still running on host [10.106.84.13]: stderr: [], stdout: []

  with module.ahoj-cluster.rancher2_cluster_sync.wait-for-cluster-is-ready,
  on modules/rancher-worker/rancher-cluster.tf line 190, in resource "rancher2_cluster_sync" "wait-for-cluster-is-ready":
 190: resource "rancher2_cluster_sync" "wait-for-cluster-is-ready" {
```

**Comment by rawmind0** [comment]
Hi @koboltmarky , is your rke cluster deploying properly?? How looks like you cluster tf definition??

**Comment by koboltmarky** [comment]
@rawmind0 yes cluster is deploying properly.

rancher-cluster.tf:
https://gist.github.com/koboltmarky/a22c9092f33f2143ef7b2c1748581f24

**Comment by koboltmarky** [comment]
@rawmind0 with 

```
rancher2 = {
      source  = "rancher/rancher2"
      version = "1.13.0"
    }
```

it works.

**Comment by rawmind0** [comment]
@koboltmarky , tf definition seems ok, beside `nodes` info that have no effect.  May be you are getting timeout at cluster creation?? What is the real cluster state when tf fails??

**Comment by koboltmarky** [comment]
@rawmind0 cluster state is ok after a while. Unfortunately our jenkins pipeline failed because of this.
Since we do not set the timeout, the default timeout of 30 minutes applies, right? The pipeline failed always between 3 ~ 5 minutes.

**Comment by gohumble** [comment]
Same issue here. 
The cluster state shows `Waiting for [file-deployer]` when terraform returns this error, but changes to ok state after a while. Also using the default timeout. 
This does not happen with every cluster created, only once in a while. 

I am going to downgrade the provider and test an older version.

**Comment by rawmind0** [comment]
I'm not able to reproduce the issue, may be caused by a race condition. `rancher2_cluster_sync` should wait until cluster condition `Updated` exists and status is `True` or `False`. This condition status should be `Unknown` in the meanwhile the cluster is transitioning. 

@koboltmarky  @gohumble could you please provide the cluster `conditions` values, got from API  https://<rancher_url>/v3/clusters/<cluster_id> , when terraform returns with error but the cluster is still provisioning??

**Comment by gohumble** [comment]
HI @rawmind0. Here we go :) 

Error was:

```
  on customer.tf line 37, in resource "rancher2_cluster_sync" "terraform-dev-cluster":
  37: resource "rancher2_cluster_sync"  "terraform-dev-cluster" {
```

Conditions from the api, right after the error was returned (Think I was a second to late, as the [file-deployer] condition is no more present.):

```json {
  "conditions": [
    {
      "status": "True",
      "type": "Pending"
    },
    {
      "lastUpdateTime": "2021-07-27T08:43:43Z",
      "status": "True",
      "type": "Provisioned"
    },
    {
      "lastUpdateTime": "2021-07-27T08:43:58Z",
      "status": "True",
      "type": "Waiting"
    },
    {
      "lastUpdateTime": "2021-07-27T08:39:02Z",
      "status": "True",
      "type": "BackingNamespaceCreated"
    },
    {
      "lastUpdateTime": "2021-07-27T08:39:02Z",
      "status": "True",
      "type": "DefaultProjectCreated"
    },
    {
      "lastUpdateTime": "2021-07-27T08:39:02Z",
      "status": "True",
      "type": "SystemProjectCreated"
    },
    {
      "lastUpdateTime": "2021-07-27T08:39:02Z",
      "status": "True",
      "type": "InitialRolesPopulated"
    },
    {
      "lastUpdateTime": "2021-07-27T08:39:02Z",
      "status": "True",
      "type": "CreatorMadeOwner"
    },
    {
      "lastUpdateTime": "2021-07-27T08:39:02Z",
      "status": "True",
      "type": "NoDiskPressure"
    },
    {
      "lastUpdateTime": "2021-07-27T08:39:02Z",
      "status": "True",
      "type": "NoMemoryPressure"
    },
    {
      "lastUpdateTime": "2021-07-27T08:43:46Z",
      "status": "False",
      "type": "AlertingEnabled"
    },
    {
      "lastUpdateTime": "2021-07-27T08:43:50Z",
      "status": "True",
      "type": "SystemAccountCreated"
    },
    {
      "lastUpdateTime": "2021-07-27T08:43:51Z",
      "status": "True",
      "type": "AgentDeployed"
    },
    {
      "lastUpdateTime": "2021-07-27T08:43:52Z",
      "status": "False",
      "type": "PrometheusOperatorDeployed"
    },
    {
      "lastUpdateTime": "2021-07-27T08:43:58Z",
      "status": "True",
      "type": "Ready"
    },
    {
      "lastUpdateTime": "2021-07-27T08:45:01Z",
      "message": "[remove/etcd-fix-perm] Successfully removed container on host [10.1.82.112]",
      "status": "False",
      "type": "Updated"
    },
    {
      "lastUpdateTime": "2021-07-27T08:44:07Z",
      "status": "True",
      "type": "ServiceAccountMigrated"
    }
  ]
}
```

as you can see, there is a `Updated=False` condition present. This condition is still changing, until the cluster is ready. 

Here are some more updated conditions after `rancher2_cluster_sync` returned the error and the cluster creation is still in progress. 

```json
{
"lastUpdateTime": "2021-07-27T08:48:06Z",
"message": "[sidekick] Sidekick container already created on host [10.1.82.113]",
"status": "False",
"type": "Updated"
}
```

```json 
{
"lastUpdateTime": "2021-07-27T08:49:51Z",
"message": "Container [etcd-fix-perm] is still running on host [10.1.82.111]: stderr: [], stdout: []",
"status": "Unknown",
"type": "Updated"
}
```

Now the cluster is ready and available. (about 4 minutes later)

```json
{
"lastUpdateTime": "2021-07-27T08:49:16Z",
"status": "True",
"type": "Updated"
},
```

Im going to try to recapture the conditions, to see if `[file-deployer]` condition is also  in status `false`. Hope this takes the investigation further. 

btw i was running this with the `state_confirm` argument set to 2 

# Edit:
Here is the file-deployer message set to status "false"
```json
{
"lastUpdateTime": "2021-07-27T09:27:50Z",
"message": "Container [file-deployer] is still running on host [10.1.82.112]: stderr: [], stdout: []",
"status": "False",
"type": "Updated"
},
```

**Comment by gohumble** [comment]
Was also digging into to provider code in the meantime. Major probs. Love the work you've done here and there is a lot to learn for me! 
I think I've found the line where the error is produced. 
https://github.com/rancher/terraform-provider-rancher2/blob/master/rancher2/config.go#L738

It seems like before this [commit](https://github.com/rancher/terraform-provider-rancher2/commit/3bd76e40da8f0951a71ba6af750b8ecd8e96abf6), the resource was waiting for the `Cluster.State` to be `active`. Now it's watching for the `Updated` condition. (correct me if im wrong).

NOTE: 
The `Cluster.State` is not `active` yet, when the error is returned. 

Possible solutions could be (just thinking)
- watching for the `Cluster.State` again 
- or implementing some kind of retry logic in https://github.com/rancher/terraform-provider-rancher2/blob/master/rancher2/config.go#L738

**Comment by gohumble** [comment]
Now I've downgraded to version `1.13.0`. Still seeing some `Updated` conditions in the api. 
`rancher_cluster_sync` resource is not returning any error, as the `Cluster.State` is still at `Updating`. 
This is what I've expected. 

```json
{
"lastUpdateTime": "2021-07-27T10:22:14Z",
"message": "[add/etcd] Successfully Added member [etcd-terraform-k8s-node02] to etcd cluster",
"status": "False",
"type": "Updated"
}
```

```json 
{
"lastUpdateTime": "2021-07-27T10:22:42Z",
"message": "Pre-pulling kubernetes images",
"status": "False",
"type": "Updated"
}
```

```json
{
"lastUpdateTime": "2021-07-27T10:23:26Z",
"message": "[etcd] Building up etcd plane..",
"status": "False",
"type": "Updated"
}
```

```json
"state": "updating"
```

**Comment by earthquakesan** [comment]
Encountered this today, rancher2 version is set to:

```
rancher2 = {
      source  = "rancher/rancher2"
      version = "~> 1.15.1"
    }
```

**Comment by rawmind0** [comment]
@gohumble @earthquakesan , what Rancher2 version are you using?? Still not able to reproduce the issue...

**Comment by gohumble** [comment]
@rawmind0 provider configuration was 
``` 
rancher2 = {
      source = "rancher/rancher2"
      version = "1.15.1"
    }
```
for me. 

Downgrad to `1.13.0` helped allot. The issue seems to be unpresent in this previous version.

**Comment by rawmind0** [comment]
@gohumble , thanks for this info, but i'm asking for Rancher version that you are using nor the tf provider.

**Comment by gohumble** [comment]
@rawmind0 sorry was not attentive here. Im running rancher `v2.5.8`

**Comment by earthquakesan** [comment]
@rawmind0 Rancher v2.5.7. The issue is not reproducible every time. We are creating ephemeral clusters all the time and the issue happened first time today (running Rancher for 2+ months).

Upd: we encountered this issue again today (10.08.2021) with another cluster.

**Comment by rawmind0** [comment]
@earthquakesan , yes, it's weird as just happening eventually. 

Submitted PR https://github.com/rancher/terraform-provider-rancher2/pull/721 adding extra check for `Cluster.Transitioning == "yes"` when the condition is false

**Comment by rawmind0** [comment]
@gohumble @earthquakesan , could you please test it?? 

Attaching binaries built from PR:
* [terraform-provider-rancher2_darwin_amd64.zip](https://github.com/rancher/terraform-provider-rancher2/files/6970042/terraform-provider-rancher2_darwin_amd64.zip)
* [terraform-provider-rancher2_linux_amd64.zip](https://github.com/rancher/terraform-provider-rancher2/files/6970043/terraform-provider-rancher2_linux_amd64.zip)
* [terraform-provider-rancher2_windows_amd64.exe.zip](https://github.com/rancher/terraform-provider-rancher2/files/6970044/terraform-provider-rancher2_windows_amd64.exe.zip)

**Comment by rawmind0** [comment]
PR #721 has been merged. Fix will be available at tfp v1.17.0

Please, reopen issue if required.

**Comment by earthquakesan** [comment]
Hi @rawmind0 ! As the issue is not reproducible all the time, we can not really test it =\ Will switch to 1.17 and see if the issue reoccurs after the switch.

**Comment by koboltmarky** [comment]
@rawmind0 we have switch to 1.17.0 and then random errros happened. Ten times it works then it failed two times.
Error message:

```
Error: Cluster ID c-tgqxj: Container [file-deployer] is still running on host [10.106.84.13]: stderr: [], stdout: []

  with module.ahoj-cluster.rancher2_cluster_sync.wait-for-cluster-is-ready,
  on modules/rancher-worker/rancher-cluster.tf line 190, in resource "rancher2_cluster_sync" "wait-for-cluster-is-ready":
 190: resource "rancher2_cluster_sync" "wait-for-cluster-is-ready" {
```

Can you reopen the ticket please ?
We don't have the ouput of the api yet, but we try to implemented that. At the moment is is really hard to debug, maybe it would make sense to output the whole object here ? https://github.com/rancher/terraform-provider-rancher2/blob/1eeba869f2b55dd119bb767928426e1aae04503a/rancher2/config.go#L742

**Comment by gohumble** [comment]
@rawmind0 I've been testing the provider version`v1.17.1` today.  The error still occurs. 

```
Error: Cluster ID c-xhzqj: Container [file-deployer] is still running on host [10.1.82.111]: stderr: [], stdout: []

  on customer.tf line 37, in resource "rancher2_cluster_sync" "foo-customs":
  37: resource "rancher2_cluster_sync"  "foo-customs" {
```

**Comment by rawmind0** [comment]
@gohumble , unfortunately this is expected due to 1.17.1 doesn't have any other fix for this. 

The issue seems a race condition occurring at Rancher not at the tf provider. The condition should be `unknown` until the transition is finished, if not there is no chance to check the cluster state from the API. Investigating on this way.

**Comment by rawmind0** [comment]
It seems the issue is caused by a cluster state flapping.

Submitted PR https://github.com/rancher/terraform-provider-rancher2/pull/732 to fix the issue. When the condition is false, added extra check if `Condition.LastUpdateTime` > 120s,  before returning error.

**Comment by rawmind0** [comment]
PR #732  has been merged. Fix is available at tfp v1.17.2

Please, reopen issue if required.
